README.md
# cellocator-parser

[![npm version](https://img.shields.io/npm/v/@drivetech/cellocator-parser.svg)](https://www.npmjs.com/package/@drivetech/cellocator-parser)
[![npm downloads](https://img.shields.io/npm/dm/@drivetech/cellocator-parser.svg)](https://www.npmjs.com/package/@drivetech/cellocator-parser)
[![Build Status](https://travis-ci.org/Drivetech/cellocator-parser.svg?branch=master)](https://travis-ci.org/Drivetech/cellocator-parser)
[![Coverage Status](https://coveralls.io/repos/github/Drivetech/cellocator-parser/badge.svg?branch=master)](https://coveralls.io/github/Drivetech/cellocator-parser?branch=master)
[![Maintainability](https://api.codeclimate.com/v1/badges/b34ffadccf17004a3dae/maintainability)](https://codeclimate.com/github/Drivetech/cellocator-parser/maintainability)
[![dependency Status](https://img.shields.io/david/drivetech/cellocator-parser.svg)](https://david-dm.org/drivetech/cellocator-parser#info=dependencies)
[![devDependency Status](https://img.shields.io/david/dev/drivetech/cellocator-parser.svg)](https://david-dm.org/drivetech/cellocator-parser#info=devDependencies)

> Parse raw data from cellocator devices

## Installation

```bash
npm i -S @drivetech/cellocator-parser
```

## Use

[Try on Tonic](https://tonicdev.com/npm/@drivetech/cellocator-parser)
```js
const cellocator = require('@drivetech/cellocator-parser');

const raw = new Buffer('4d43475000bdda0b0000060ddf20041017002000e3c40000baeff3c6b6224502000000000000ea65000402090daec5f7cb302cff3357000038090000930a002a170c03e007c1', 'hex');
const data = cellocator.parse(raw);
/*{
  raw: '4d43475000bdda0b0000060ddf20041017002000e3c40000baeff3c6b6224502000000000000ea65000402090daec5f7cb302cff3357000038090000930a002a170c03e007c1',
  unitId: 776893,
  imei: 0,
  manufacturer: 'cellocator',
  device: 'CelloTrack',
  alarm: {type: 'ConnectionUp'},
  type: 'data',
  loc: {
    type: 'Point',
    coordinates: [ -79.09097658351084, -7.953307941260071 ]
  },
  speed: 84.96,
  datetime: '2016-03-12T23:42:00.000Z',
  gpsTime: '2016-07-12T23:42:00.000Z',
  direction: 155.09967514191385,
  satellites: 9,
  voltage: {
    ada: 28.1176470588165,
    adb: 4.00235293989,
    adc: 45.41720000000001,
    add: 182
  },
  altitude: 223.23000000000002,
  status: {
    sos: false,
    input: {
      '1': false,
      '2': false,
      '3': false,
      '4': false,
      '5': true
    },
    output: {
      '1': false,
      '2': false,
      '3': false,
      '4': false,
      '5': false,
      '6': false
    },
    charge: false,
    engine: false,
    driving: false,
    accelerometer: true,
    gpsPower: false
  },
  communication: {
    noHibernation: false,
    momentarySpeed: false,
    privateMode: false,
    firmwareSubversion: '0011',
    canOriginatedOdometer: false,
    canOriginatedSpeed: false,
    dataType33_38: '0',
    messageSource: '0',
    garminConnected: false,
    garminEnable: false,
    messageInitiative: false
  },
  locationStatus: {
    GNSSAntennaSelected: 'internal',
    trailer: false,
    CFEType: 'Not Applicable'
  },
  hardware: {
    id: 31,
    model: 'Cello-IQ',
    modem: 'Telit GE864, automative'
  },
  software: 32,
  protocol: 4,
  transmissionReason: 32,
  transmissionReasonSpecificData: 0,
  odometer: 148770,
  gpsModes: {
    '1': {
      pMode: '>3 satellite solution',
      tpMode: 'Full power position',
      altMode: 'No altitude hold',
      dopMask: 'DOP mask not exceeded',
      dgps: 'No DGPS position'
    },
    '2': 'Validated'
  },
  plmn: 71610,
  serialId: 13,
  messageType: 0,
  valid: true
}*/

const data = {
  unitId: 836522,
  commandNumerator: 1,
  instruction: '1_on' // enable output 1 => port + _ + on/off
};
const command = cellocator.parseCommand(data);
// new Buffer('4d43475000aac30c0001000000000003000300140014000000000000a8', 'hex')
```

## License

[MIT](https://tldrlegal.com/license/mit-license)


CHANGELOG.md
#### 2.2.10 (2019-01-02)

##### Chores

* **deps:**
  *  update dependency chai to v4.2.0 ([36bd3091](https://github.com/drivetech/cellocator-parser/commit/36bd3091a405a7cbc91dd3ad8bafc7410e31acea))
  *  update dependency codeclimate-test-reporter to v0.5.1 ([cfe8833a](https://github.com/drivetech/cellocator-parser/commit/cfe8833aea16a930e932b046f412ca17548ac77f))
  *  update dependency eslint to v5.11.0 ([ed31da64](https://github.com/drivetech/cellocator-parser/commit/ed31da64770c7281399c9b76382756d4b6d90a34))
  *  update dependency eslint to v5 ([a7d1eebc](https://github.com/drivetech/cellocator-parser/commit/a7d1eebcf2ffd62d3f1970fa2114602763139253))
  *  update dependency coveralls to v3.0.2 ([5f4b9402](https://github.com/drivetech/cellocator-parser/commit/5f4b94022912889f744c31abfa450be10c66d515))
  *  update dependency mocha to v5.2.0 ([4f916dbe](https://github.com/drivetech/cellocator-parser/commit/4f916dbed2f5a211b5983ae1b47a12e4b8cb6677))
  *  update dependency coveralls to v3.0.1 ([6c2432bd](https://github.com/drivetech/cellocator-parser/commit/6c2432bda48376ed3c5aa3ad64f59679f258f466))
  *  update dependency mocha to v5.1.1 ([c2cea6a4](https://github.com/drivetech/cellocator-parser/commit/c2cea6a48ee1a4f67b1e9073571da4ba92c7a266))
  *  update dependency mocha to v5.1.0 ([990ad2a0](https://github.com/drivetech/cellocator-parser/commit/990ad2a0130f8365557afd5ce4187907163c4255))
  *  update dependency generate-changelog to v1.7.1 ([90185cc8](https://github.com/drivetech/cellocator-parser/commit/90185cc83c57e4cb770a2414eb5a6f24362fc3b9))
  *  update dependency mocha to v5.0.5 ([61711089](https://github.com/drivetech/cellocator-parser/commit/61711089969ca620bcc2e9c2920cc9b9e51e1392))
  *  update dependency eslint to v4.19.1 ([a30d4c51](https://github.com/drivetech/cellocator-parser/commit/a30d4c516c298f1b42e267f50bc73c39c9d24595))
  *  update dependency eslint to v4.19.0 ([fd2d7404](https://github.com/drivetech/cellocator-parser/commit/fd2d74044b8d77b34c1fe2801f5447ffa0f56f36))
  *  update dependency mocha to v5.0.4 ([2bd39ec4](https://github.com/drivetech/cellocator-parser/commit/2bd39ec45d586e656932c18f5c40f98a991cb83e))
  *  update dependency mocha to v5.0.3 ([4903ead8](https://github.com/drivetech/cellocator-parser/commit/4903ead8543503c072cf6da7927c52c4dd3c6c52))
  *  update dependency mocha to v5.0.2 ([09f82ab8](https://github.com/drivetech/cellocator-parser/commit/09f82ab83cf14f1a4bc09df7db547e4767482842))
  *  update dependency eslint to v4.18.2 ([19a12172](https://github.com/drivetech/cellocator-parser/commit/19a121729d674675b3bd8b36628ee70c766baad9))
  *  update dependency eslint to v4.18.1 ([a69b72e0](https://github.com/drivetech/cellocator-parser/commit/a69b72e0743050c6e36cd2486f389ac2a3178a89))
  *  update dependency eslint to v4.18.0 ([30f37bd2](https://github.com/drivetech/cellocator-parser/commit/30f37bd2d380b19adcea8ea67ee8c42533a849f3))
  *  update dependency mocha to v5.0.1 ([49658309](https://github.com/drivetech/cellocator-parser/commit/4965830954684f9ff2ad66a8e0686781ca98f51c))
  *  update dependency eslint to v4.17.0 ([eb5c18f5](https://github.com/drivetech/cellocator-parser/commit/eb5c18f53b5dfa109fe1ec4bb3e4280e79a0d27f))
  *  update dependency eslint to v4.16.0 ([bec0c605](https://github.com/drivetech/cellocator-parser/commit/bec0c6051816d21e5f1678376c1bfe8c82267e1a))
  *  update dependency mocha to v5.0.0 ([c3fadcef](https://github.com/drivetech/cellocator-parser/commit/c3fadcef3c94898d7c7b3a56856da1fa386a8603))
  *  update dependency mocha to v4.1.0 ([f15733b6](https://github.com/drivetech/cellocator-parser/commit/f15733b6b38c2e14b954b33b55c56b248b35e60a))
  *  update dependency eslint to v4.14.0 ([a9bd702d](https://github.com/drivetech/cellocator-parser/commit/a9bd702d1ac6d9783471c2ce6dcb101ec8aeb408))
  *  update dependency generate-changelog to v1.7.0 ([4df93743](https://github.com/drivetech/cellocator-parser/commit/4df9374319d55d4223e19e20f9d9bc4dca54c48c))
  *  update dependency eslint to v4.13.1 ([5b8bf0dd](https://github.com/drivetech/cellocator-parser/commit/5b8bf0dd44677c0d1b1ca439dc78a5e0a6ee2428))
  *  update dependency eslint to v4.13.0 ([fbf684fb](https://github.com/drivetech/cellocator-parser/commit/fbf684fbfae72f6cc7194a3393a094d47a8452a6))
  *  update dependency eslint to v4.12.1 ([0f1dcbe1](https://github.com/drivetech/cellocator-parser/commit/0f1dcbe1e81302a63c805a2deb23f3bff5a691e1))
  *  update dependency eslint to v4.12.0 ([6f55af26](https://github.com/drivetech/cellocator-parser/commit/6f55af269bd0bcea6956ff3cfa16ac321e2444f2))
  *  update dependency generate-changelog to v1.6.0 ([c0d7f4a1](https://github.com/drivetech/cellocator-parser/commit/c0d7f4a1c308e0d68aa4095267e5b56bb676e07d))
  *  update dependency eslint to v4.11.0 ([d16128e4](https://github.com/drivetech/cellocator-parser/commit/d16128e4095e42a44d2412e16d77640b457b7b26))
  *  update dependency eslint to v4.10.0 ([b87d40fd](https://github.com/drivetech/cellocator-parser/commit/b87d40fdd4c03cc7ac05fe2aaeb8a3188787dd2d))
  *  update dependency eslint to v4.9.0 ([b87d2cc0](https://github.com/drivetech/cellocator-parser/commit/b87d2cc0a60cd92099e6db774960206fb4b25ddc))
  *  update dependency mocha to v4.0.1 ([d9a5a16e](https://github.com/drivetech/cellocator-parser/commit/d9a5a16ec92b30c08bad56365984cee0ddd7c710))
  *  update dependency mocha to v4.0.0 ([bd6c78fe](https://github.com/drivetech/cellocator-parser/commit/bd6c78fe62cd2a3b15196ab34d8072d3a89f0a62))
  *  update dependency eslint to v4.8.0 ([00b8084b](https://github.com/drivetech/cellocator-parser/commit/00b8084bf66a870abc944c5acc45adea6562e0c7))
  *  update dependency coveralls to v3.0.0 ([94152f52](https://github.com/drivetech/cellocator-parser/commit/94152f523dfe327d315c0d3c1deaa6d91cfcdaca))
  *  update dependency coveralls to v2.13.3 ([9b2d6082](https://github.com/drivetech/cellocator-parser/commit/9b2d60823160589a29a064efcd79136fb64c1058))

src/index.js
'use strict';

const utils = require('./utils');
const inbound = require('./inbound');

const storeIncompleteData = {};

const parse = raw => {
  const rawString = raw.toString('hex');
  if (!utils.patterns.data.test(rawString)) return {type: 'UNKNOWN', raw: rawString};
  const results = rawString.split(utils.patterns.multi).filter(x => utils.patterns.data.test(x)).map(utils.getData);
  const incompleteData = rawString.split(utils.patterns.multi).filter(x => !utils.patterns.data.test(x) && x !== '');
  if (incompleteData.length > 0) {
    const reply = storeIncompleteData[results[0].unitId];
    if (reply) {
      let previousData = '';
      if (/^4d434750/.test(reply)) {
        previousData = new Buffer(`${reply}${incompleteData[0]}`, 'hex');
      } else if (/^4d434750/.test(incompleteData[0])) {
        previousData = new Buffer(`${incompleteData[0]}${reply}`, 'hex');
      }
      if (utils.patterns.data.test(previousData.toString('hex'))) {
        results.unshift(utils.getData(previousData));
      }
    } else {
      storeIncompleteData[results[0].unitId] = incompleteData[0];
    }
  }
  if (results.length === 1) return results[0];
  return results;
};

const isCello = raw => utils.patterns.data.test(raw.toString('hex'));

module.exports = {
  parse: parse,
  patterns: utils.patterns,
  isCello: isCello,
  getImei: utils.getImei,
  ack: inbound.ack,
  parseCommand: inbound.parseCommand
};


src/inbound.js
'use strict';

const core = require('./core');
const utils = require('./utils');
const commands = require('./commands');

const genericCommand = (uId, commandNumeratorField, codefield, firsttwice) => {
  const systemCode = utils.ascii2hex('MCGP');
  const messageType = '00';
  // unitId to hex with padding of 4 bytes and reverse
  const unitId = utils.reverseHex(core.lpad(core.convertBase(uId, 10, 16), 8));
  // commandNumerator to hex
  const commandNumerator = core.lpad(core.convertBase(commandNumeratorField > 255 ? 0 : commandNumeratorField, 10, 16), 2);
  const authenticationCode = '00000000';
  const commandCodeField = core.lpad(codefield, 4);
  const firstCommandDataField = core.lpad(firsttwice, 4);
  const secondCommandDataField = '0000';
  const commandSpecificDataField = '00000000';
  const trama = `${messageType}${unitId}${commandNumerator}${authenticationCode}${commandCodeField}${commandCodeField}${firstCommandDataField}${firstCommandDataField}${secondCommandDataField}${commandSpecificDataField}`;
  const checksum = utils.checksum(trama);
  const command = `${systemCode}${trama}${checksum}`;
  return new Buffer(command, 'hex');
};

exports.ack = (uId, commandNumerator, messageNumerator) => {
  const systemCode = utils.ascii2hex('MCGP');
  const messageType = '04';
  // unitId to hex with padding of 4 bytes and reverse
  const unitId = utils.reverseHex(core.lpad(core.convertBase(uId, 10, 16), 8));
  // commandNumerator to hex
  const commandNumeratorField = core.lpad(core.convertBase(commandNumerator > 255 ? 0 : commandNumerator, 10, 16), 2);
  const authenticationCodeField = '00000000';
  const actionCode = '00';
  const mainAckNumberLSB = core.lpad(core.convertBase(messageNumerator, 10, 16), 2);
  const mainAckNumberMSB = '00';
  const secundaryAckNumberLSB = '00';
  const secundaryAckNumberMSB = '00';
  const reserve1 = '00';
  const compressDate = '0000';
  const compressTime = '000000';
  const reserve2 = '0000';
  const trama = `${messageType}${unitId}${commandNumeratorField}${authenticationCodeField}${actionCode}${mainAckNumberLSB}${mainAckNumberMSB}${secundaryAckNumberLSB}${secundaryAckNumberMSB}${reserve1}${compressDate}${compressTime}${reserve2}`;
  const checksum = utils.checksum(trama);
  const command = `${systemCode}${trama}${checksum}`;
  return new Buffer(command, 'hex');
};

const parseIOCommand = instruction => {
  const outputs = {
    '1': {off: '04', on: '14'},
    '2': {off: '00', on: '10'},
    '3': {off: '05', on: '15'},
    '4': {off: '03', on: '13'},
    '5': {off: '08', on: '18'},
    '6': {off: '09', on: '19'}
  };
  let data = instruction.split('_');
  const port = data[0];
  const state = data[1];
  return {codeField: '03', dataField: outputs[port][state]};
};

const instructionParsers = data => {
  let result = null;
  const parsers = [
    {regex: /^[1-5]{1}_(on|off)$/, parser: parseIOCommand},
  ];
  const parser = parsers.find(x => x.regex.test(data.instruction));
  if (typeof parser === 'undefined') {
    const command = commands[data.type];
    if (typeof command !== 'undefined') result = command;
  } else {
    result = parser.parser(data.instruction);
  }
  return result;
};

exports.parseCommand = data => {
  const command = instructionParsers(data);
  return genericCommand(data.unitId, data.commandNumerator, command.codeField, command.dataField);
};


src/utils.js
'use strict';

const hardwares = require('./hardwares');
const core = require('./core');
const parseAlarm = require('./alarms');
const isValid = require('date-fns/is_valid');
const format = require('date-fns/format');

const patterns = {
  multi: /(4d434750[A-F0-9]{132})/gi,
  data: /(4d434750[A-F0-9]{132})/i
};

const ascii2hex = data => data.split('').map(x => x.charCodeAt(0).toString(16).toUpperCase()).join('');

const reverseHex = data => {
  data = data.toUpperCase();
  if (data.length % 2 !== 0) data = `0${data}`;
  return data.split(/([A-F0-9]{2})/).filter(x => x !== '').reverse().join('');
};

const parseMessageType = byte5 => core.convertBase(byte5, 16, 10);

const parseUnitsId = bytes6To9 => core.convertBase(reverseHex(bytes6To9), 16, 10);

const parseCommunication = bytes10To11 => {
  const byte10 = core.hex2bin(bytes10To11.substring(0, 2));
  const byte11 = core.hex2bin(bytes10To11.substring(2, 4));
  return {
    noHibernation: byte11[0] === '1',
    momentarySpeed: byte11[1] === '1',
    privateMode: byte11[2] === '1',
    firmwareSubversion: byte11.substring(3, 7),
    canOriginatedOdometer: byte10[0] === '1',
    canOriginatedSpeed: byte10[1] === '1',
    dataType33_38: byte10.substring(2, 3),
    messageSource: byte10[4],
    garminConnected: byte10[5] === '1',
    garminEnable: byte10[6] === '1',
    messageInitiative: byte10[7] === '1'
  };
};

const parseMessageNumerator = byte12 => core.convertBase(byte12, 16, 10);

const parseHardware = byte13 => {
  const data = core.hex2bin(byte13);
  const modem  = core.convertBase(data.substr(0, 3), 2, 10);
  const model = core.convertBase(data.substr(3, 5), 2, 10);
  const hardware = hardwares.find(x => x.model.id === model && x.modem.code === modem);
  return {
    id: hardware.model.id,
    model: hardware.model.name,
    modem: hardware.modem.name
  };
};

const parseSoftware = byte14 => {
  return core.convertBase(byte14, 16, 10);
};

const protocolVersion = byte15 => core.convertBase(byte15, 16, 10);

const parseUnitsStatusCurrentGsmOperator = byte16 => {
  const firstnibble = byte16[0];
  const unitsstatus = core.convertBase(byte16[1], 16, 2);
  return {
    speedEstimatedByGps: unitsstatus[0] === '1',
    correctTime: unitsstatus[1] === '1',
    homeNetwork: unitsstatus[2] === '1',
    gpsCommunicationAvailable: unitsstatus[3] === '1',
    currentGsmOperator1stNibble: firstnibble
  };
};

const parseCurrentGsmOperator = byte17 => byte17;

const parseTransmissionReasonSpecificData = byte18 => core.convertBase(byte18, 16, 10);

const parseTransmissionReason = byte19 => core.convertBase(byte19, 16, 10);

const parseEngineStatus = byte20 => core.convertBase(byte20, 16, 10) === 1;

const parseIO = bytes21To24 => {
  const byte21 = core.hex2bin(bytes21To24.substring(0, 2));
  const byte22 = core.hex2bin(bytes21To24.substring(2, 4));
  const byte23 = core.hex2bin(bytes21To24.substring(4, 6));
  const byte24 = core.hex2bin(bytes21To24.substring(6, 8));
  return {
    unlock: byte21[0] === '0',
    sos: byte21[1] === '0',
    driving: byte21[2] === '0',
    shock: byte21[6] === '0',
    door: byte21[7] === '0',
    ignition: byte22[0] === '1',
    accelerometer: byte22[1] === '1',
    lock: byte22[5] === '0',
    gpsPower: byte23[4] === '1',
    gradualStop: byte23[5] === '1',
    siren: byte23[6] === '1',
    cfeOut1: byte23[7] === '1',
    charge: byte24[0] === '1',
    standardImmobilizer: byte24[2] === '1',
    blinker: byte24[4] === '1',
    ledOut: byte24[7] === '1'
  };
};

const parsePlmn = (byte25, currentGsmOperator1stNibble, currentGsmOperator) => {
  return core.convertBase(`${currentGsmOperator1stNibble}${currentGsmOperator}${byte25}`, 16, 10);
};

const parseAnalogInput1 = byte26 => core.convertBase(byte26, 16, 10) * 0.1176470588235;

const parseAnalogInput2 = byte27 => core.convertBase(byte27, 16, 10) * 0.01647058823;

const parseAnalogInput3 = byte28 => (core.convertBase(byte28, 16, 10) * 0.4314) - 40;

const parseAnalogInput4 = byte29 => core.convertBase(byte29, 16, 10);

const parseMileageCounter = bytes30To32 => core.convertBase(reverseHex(bytes30To32), 16, 10);

const multiPurposeField = (bytes33To38, byte41) => {
  const ms4And5From41 = core.hex2bin(byte41).substring(1, 3);
  return parseInt(`${ms4And5From41}${core.hex2bin(reverseHex(bytes33To38))}`, 2);
};

const parseGpsTime = (bytefrom39to40, seconds) => {
  seconds = core.lpad(core.convertBase(seconds, 16, 10), 2);
  const now = format(new Date(), 'YYYY-MM');
  const data = core.hex2bin(reverseHex(bytefrom39to40));
  const day = `00${core.convertBase(data.substr(0, 5), 2, 10)}`.substr(-2);
  const hours = `00${core.convertBase(data.substr(5, 5), 2, 10)}`.substr(-2);
  const minutes = `00${core.convertBase(data.substr(10, 6), 2, 10)}`.substr(-2);
  const datetime = new Date(`${now}-${day}T${hours}:${minutes}:${seconds}`);
  return isValid(datetime) ? datetime.toISOString() : null;
};

const parseLocationStatus = byte41 => {
  const data = core.hex2bin(byte41);
  const CFETypes = {
    '000': 'Not Applicable',
    '001': 'CFE is not connected',
    '010': 'CFE BT is connected',
    '011': 'CFE Basic is connected',
    '100': 'CFE I/O is connected',
    '101': 'CFE premium is connected',
    '111': 'Undefined CFE Type'
  };
  return {
    GNSSAntennaSelected: data[7] === '0' ? 'internal' : 'external',
    trailer: data[6] === '1',
    CFEType: CFETypes[data.substring(3, 6)]
  };
};

/*
With Normal PMODE Filter settings, PMODE valid is 2 to 6
With Tide PMODE Filter settings PMODE valid 3 or 4
*/
const parseMode1 = byte42 => {
  const pModes = {
    '0': 'No navigation solution',
    '1': '1 satellite solution',
    '2': '2 satellite solution',
    '3': '3 satellite solution',
    '4': '>3 satellite solution',
    '5': '2D point solution (Least square)',
    '6': '3D point solution (Least square)',
    '7': 'Dead reckoning'
  };
  const tpModes = {
    '0': 'Full power position',
    '1': 'Trickle power position'
  };
  const altModes = {
    '0': 'No altitude hold',
    '1': 'Altitude used from filter',
    '2': 'Altitude used from user',
    '3': 'Force altitude (from user)'
  };
  const dopMasks = {
    '0': 'DOP mask not exceeded',
    '1': 'DOP mask exceeded'
  };
  const dgpss = {
    '0': 'No DGPS position',
    '1': 'DGPS position'
  };
  const modes = /([01]{1})([01]{1})([01]{2})([01]{1})([01]{3})/i.exec(core.hex2bin(byte42));
  const dgps = core.convertBase(modes[1], 2, 10);
  const dopMask = core.convertBase(modes[2], 2, 10);
  const altMode = core.convertBase(modes[3], 2, 10);
  const tpMode = core.convertBase(modes[4], 2, 10);
  const pMode = core.convertBase(modes[5], 2, 10);
  return {
    pMode: pModes[pMode],
    tpMode: tpModes[tpMode],
    altMode: altModes[altMode],
    dopMask: dopMasks[dopMask],
    dgps: dgpss[dgps]
  };
};

const parseMode2 = byte43 => {
  const modes = {
    '00': 'Solution not validated',
    '01': 'DR Sensor Data',
    '02': 'Validated',
    '04': 'Dead Reckoning',
    '08': 'Output edited by UI',
    '10': 'Reserved',
    '20': 'Reserved',
    '40': 'Reserved',
    '80': 'Reserved'
  };
  return byte43 in modes ? modes[byte43] : null;
};

const parseSatellites = byte44 => core.convertBase(byte44, 16, 10);

const notBinaryData = data => data.split('').map(x => x === '1' ? '0' : '1').join('');

const parseLatitude = bytes45To48 => {
  let lat = 0;
  const data = reverseHex(bytes45To48);
  if (data[0] === 'F') {
    lat = core.convertBase(notBinaryData(core.hex2bin(data)), 2, 10);
    lat = (lat + 1) * -1;
  } else {
    lat = core.convertBase(data, 16, 10);
  }
  return lat * (180 / Math.PI) * Math.pow(10, -8);
};

const parseLongitude = bytes49To52 => {
  let lng = 0;
  const data = reverseHex(bytes49To52);
  if (data[0] === 'F') {
    lng = core.convertBase(notBinaryData(core.hex2bin(data)), 2, 10);
    lng = (lng + 1) * -1 * (180 / Math.PI) * Math.pow(10, -8);
  } else {
    lng = core.convertBase(data, 16, 10);
    lng = lng * (180 / Math.PI) * Math.pow(10, -9);
  }
  return lng;
};

const parseLoc = (bytes45To48, bytes49To52) => {
  return {
    type: 'Point',
    coordinates: [parseLongitude(bytes45To48), parseLatitude(bytes49To52)]
  };
};

const parseAltitude = bytes53To56 => core.convertBase(reverseHex(bytes53To56), 16, 10) * 0.01;

const parseSpeed = bytes57To60 => core.convertBase(reverseHex(bytes57To60), 16, 10) * 0.036;

const parseDirection = bytes61To62 => core.convertBase(reverseHex(bytes61To62), 16, 10) * (180 / Math.PI) * 0.001;

const parseDatetime = (bytes68To69, byte67, byte66, byte65, byte64, byte63) => {
  const year = core.lpad(core.convertBase(reverseHex(bytes68To69), 16, 10), 2);
  const month = core.lpad(core.convertBase(byte67, 16, 10), 2);
  const day = core.lpad(core.convertBase(byte66, 16, 10), 2);
  const hour = core.lpad(core.convertBase(byte65, 16, 10), 2);
  const minute = core.lpad(core.convertBase(byte64, 16, 10), 2);
  const second = core.lpad(core.convertBase(byte63, 16, 10), 2);
  const datetime = new Date(`${year}-${month}-${day}T${hour}:${minute}:${second}+00:00`);
  return isValid(datetime) ? datetime.toISOString() : null;
};

const checksum = trama => {
  const code = trama.split(/([A-F0-9]{2})/i).filter(x => x !== '').reduce((prev, curr) => prev + core.convertBase(curr, 16, 10), 0);
  return `00${core.convertBase(code, 10, 16)}`.toUpperCase().substr(-2);
};

const validate = (byte70, bytes4To69) => {
  return checksum(bytes4To69) === byte70.toUpperCase();
};

const getData = raw => {
  const bytes = raw.toString('hex').split(/([A-F0-9]{2})/i).filter(x => x !== '');
  const seconds = bytes[62];
  const io = parseIO(bytes.slice(20, 24).join(''));
  const unitsStatusCurrentGsmOperator = parseUnitsStatusCurrentGsmOperator(bytes[15]);
  const currentGsmOperator = parseCurrentGsmOperator(bytes[16]);
  return {
    raw: bytes.join(''),
    unitId: parseUnitsId(bytes.slice(5, 9).join('')),
    imei: multiPurposeField(bytes.slice(32, 38).join(''), bytes[40]),
    manufacturer: 'cellocator',
    device: 'CelloTrack',
    type: 'data',
    alarm: parseAlarm(bytes[18], bytes[17]),
    loc: parseLoc(bytes.slice(44, 48).join(''), bytes.slice(48, 52).join('')),
    speed: parseSpeed(bytes.slice(56, 60).join('')),
    datetime: parseDatetime(bytes.slice(67, 69).join(''), bytes[66], bytes[65], bytes[64], bytes[63], bytes[62]),
    gpsTime: parseGpsTime(bytes.slice(38, 40).join(''), seconds),
    direction: parseDirection(bytes.slice(60, 62).join('')),
    satellites: parseSatellites(bytes[43]),
    voltage: {
      ada: parseAnalogInput1(bytes[25]),
      adb: parseAnalogInput2(bytes[26]),
      adc: parseAnalogInput3(bytes[27]),
      add: parseAnalogInput4(bytes[28])
    },
    altitude: parseAltitude(bytes.slice(52, 56).join('')),
    status: {
      sos: io.sos,
      input: {
        '1': io.door,
        '2': io.shock,
        '3': io.unlock,
        '4': io.lock,
        '5': io.ignition
      },
      output: {
        '1': io.blinker,
        '2': io.siren,
        '3': io.standardImmobilizer,
        '4': io.gradualStop,
        '5': io.ledOut,
        '6': io.cfeOut1
      },
      charge: io.charge,
      engine: parseEngineStatus(bytes[19]),
      driving: io.driving,
      accelerometer: io.accelerometer,
      gpsPower: io.gpsPower
    },
    communication: parseCommunication(bytes.slice(9, 11).join('')),
    locationStatus: parseLocationStatus(bytes[40]),
    hardware: parseHardware(bytes[12]),
    software: parseSoftware(bytes[13]),
    protocol: protocolVersion(bytes[14]),
    transmissionReason: parseTransmissionReason(bytes[18]),
    transmissionReasonSpecificData: parseTransmissionReasonSpecificData(bytes[17]),
    odometer: parseMileageCounter(bytes.slice(29, 32).join('')),
    gpsModes: {
      '1': parseMode1(bytes[41]),
      '2': parseMode2(bytes[42])
    },
    plmn: parsePlmn(bytes[24], unitsStatusCurrentGsmOperator.currentGsmOperator1stNibble, currentGsmOperator),
    serialId: parseMessageNumerator(bytes[11]),
    messageType: parseMessageType(bytes[4]),
    valid: validate(bytes[69], bytes.slice(4, 69).join(''))
  };
};

const getImei = raw => {
  let imei = null;
  const data = raw.toString('hex');
  if (patterns.data.test(data)) {
    const bytes = data.split(/([A-F0-9]{2})/i).filter(x => x !== '');
    imei = multiPurposeField(bytes.slice(32, 38).join(''), bytes[40]);
  }
  return imei ? imei.toString() : null;
};

module.exports = {
  patterns: patterns,
  ascii2hex: ascii2hex,
  reverseHex: reverseHex,
  checksum: checksum,
  getData: getData,
  getImei: getImei,
};
